# Neural Network Application: Predicting SP500

In this exercise, you will build a neural network that aims to predict whether the SP500 index will go up or down based on the daily returns in the previous five days. 

###  We first import python libraries: NumPy and Pandas.

"NumPy is the fundamental package for scientific computing with Python."

"Pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool, built on top of the Python programming language."

In [1]:
import numpy as np
import pandas as pd

### We then import Keras models: "Sequential" and "Dense".

The "Sequential" model is a linear stack of layers.

The "Dense" class use fully connected layers. In a fully connected layer, each neuron receives input from every element of the previous layer.

In [2]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [3]:
from google.colab import drive 
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### Now we import the dataset "SP500dailyinputdata.csv". 
This dataset includes a variable that indicates whether the S&P500 index goes up or down in a given day and the daily returns in the previous 5 days. The sample period is from January 2015 to December 2019. 

### YOUR TURN
To import "SP500dailyinputdata.csv", complete the next line by replacing "?".

In [4]:
dataset = pd.read_csv("/content/drive/My Drive/SP500dailyinputdata.csv")

Variable description:

UpDown = indicates whether the S&P index goes up or down in day t. (0: down, 1: up)

lag1return = the return of the S&P index in t-1.

lag2return = the return of the S&P index in t-2.

lag3return = the return of the S&P index in t-3.

lag4return = the return of the S&P index in t-4.

lag5return = the return of the S&P index in t-5.


### We generate two datasets X and y from the original dataset.  

X is the dataset that contains the independent (predictive) variables. 

y is the dataset that contains the outcome variable "UpDown". 

### YOUR TURN
To generate two datasets X and y, complete the next line by replacing "?".

In [7]:
X = dataset.drop(['UpDown'],axis=1)
y = dataset['UpDown']

### Now we define the neural network.  

In [11]:
model = Sequential()

## YOUR TURN

To add a layer with 8 neurons to your neural network, complete the next line by replacing "?".

In [12]:
model.add(Dense(8, input_dim=5, activation='relu'))

To add another layer with 6 neurons to your neural network, complete the next line by replacing "?".

In [13]:
model.add(Dense(6, activation='relu'))

In [14]:
model.add(Dense(4, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

In [15]:
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

"input_dim" is the number of variables in X dataset.

Two activation functions used here include  "ReLu" and "sigmoid". We have learned "sigmoid" in our class.  

If you are interested in learning "ReLu", here is a good website:

https://www.kaggle.com/dansbecker/rectified-linear-units-relu-in-deep-learning

("ReLu" is not required by this class). 

A sigmoid works well for a classifier. ReLu is less computationally expensive than sigmoid because it involves simpler mathematical operations.

The loss function is "binary_crossentropy".  The optimizer is "Adam".

### Fit the neural network model on the dataset

In [16]:
model.fit(X, y, epochs=150, verbose=0)

epochs: Number of epochs to train the model. An epoch is an iteration over the entire x and y data provided.

verbose: display option. 0 = silent, 1 = progress bar, 2 = one line per epoch.

### Evaluate the neural network model

In [17]:
loss, accuracy = model.evaluate(X, y)

40/40 [==============================] - 0s 961us/step - loss: 0.6861 - accuracy: 0.5388


In [18]:
print(accuracy)

0.538769006729126


### YOUR TURN
Just double click the text below and you will be able to fill in the blanks. 

The accuracy of the predictions is **0.539**.  This means that **53.9 %** of cases are predicted correctly. 

Is market efficient? (you don't have to answer it here. Just give some thoughts...). 

### Congratulations! Now you have successfully built a neural network model!